Adds option to filter historical data by dividends and stock splits

hackeresq opened the pull request:
It appears that the historical data method does not include dividend or split data in the (I'm unsure whether it used to be included by Yahoo and only recently broke, or if this was an intentional omission).

To solve this, this PR adds functionality to the `getHistoricalData()` method and allows for an optional `$filter` parameter. 

This filter can be one of 'history' (default), 'div,' or 'split' which is already provided by the Yahoo Finance endpoint.

You can use this new filter by passing one of the new constants (ApiClient::FILTER_DIVIDENDS) or a simple string ('div'):

```php
$dividendData = $client->getHistoricalData("AAPL", ApiClient::INTERVAL_1_DAY, new \DateTime("-14 days"), new \DateTime("today"), ApiClient::FILTER_DIVIDENDS);
```

I've updated the readme to give an example of how this additional `$filter` parameter is **optional** and has a sensible default ('history'), such that this will not be a breaking change.  

I am unfamiliar with the testing suite in the library, and I am unclear how the mock data can be updated to reflect this new filter option, but I believe there should already be coverage for this small addition.

hackeresq committed:
adds filter to gethistoric data

hackeresq committed:
Update README.md

hackeresq committed:
clarify optional filter

hackeresq committed:
fix wrong method name

hackeresq committed:
update typing on splits

hackeresq committed:
update splits to camelcase

hackeresq committed:
trying to please Coding Standards

hackeresq committed:
trim whitespace

scheb commented:
Getting dividend data and stock splits has never been a feature of this library. I guess this is something Yahoo Finance has added at some point, but has never been added to the library. But I like it, seems useful and easy to implement :)

On your proposed changes, I'd prefer having two new method `getHistoricalDividendData` and `getHistoricalStockSplitData`. It would be a more cleaner implementation to separate the use-cases with a method each. Also, the proposed solution is somewhat breaking the method contract as it's no longer returning "array of `HistoricalData`" (although this shouldn't be a problem for backwards compatibility, as the default value would still behave as before).

In that context it would also make sense to add a `getHistoricalQuoteData` method which is identical to `getHistoricalData` method and deprecate the `getHistoricalData` method.

The part to execute the query should be a private method then, one that all of these public methods are using to retrieve their raw data.

hackeresq committed:
separate filters into separate methods and

separate filters into separate methods and depreciate the original getHistoricalData() method with friendly message in logs

hackeresq commented:
Definitely thought about having each method be separate at first, but I didn't want to have any breaking changes on my first PR ! A depreciation notice is polite enough. 

Let me know if this works?

hackeresq committed:
trying to please code sniffer

scheb commented on src/ApiClient.php:
```diff
@@ -52,35 +57,74 @@ public function search(string $searchTerm): array
-     * Get historical data for a symbol.
+     * Get historical data for a symbol (depreciated).
+     *
+     * DEPRECIATION NOTICE:
+     * In future versions, this function will be removed.
+     * Please consider using getHistoricalQuoteData() instead.
```
Would you please replace with block with `@deprecated In future versions, this function will be removed. Please consider using getHistoricalQuoteData() instead.` (in a single line). That will allow IDEs to detect the deprecation message and highlight it in the code.

Will do a proper review later today.

hackeresq replied:
<img width="631" alt="Screen Shot 2021-02-26 at 9 33 05 AM" src="https://user-images.githubusercontent.com/16628119/109320571-c33af780-7815-11eb-84e7-32f6d6f6a9f0.png">

Oh wow, fancy. I didn't even know that was a feature.

hackeresq committed:
add depreciation docblock

scheb commented on src/ApiClient.php:
```diff
@@ -16,6 +18,9 @@ class ApiClient
     public const INTERVAL_1_DAY = '1d';
     public const INTERVAL_1_WEEK = '1wk';
     public const INTERVAL_1_MONTH = '1mo';
+    public const FILTER_HISTORICAL = 'history';
+    public const FILTER_DIVIDENDS = 'div';
+    public const FILTER_SPLITS = 'split';
```
I believe we can make these private now :)

scheb commented on src/ApiClient.php:
```diff
@@ -52,35 +57,72 @@ public function search(string $searchTerm): array
-        if ($startDate > $endDate) {
-            throw new \InvalidArgumentException('Start date must be before end date');
-        }
+        error_log('[scheb/yahoo-finance-api] getHistoricalData() is deprecated and will be removed in a future release');
```
Let's use this:
```php
@trigger_error('[scheb/yahoo-finance-api] getHistoricalData() is deprecated and will be removed in a future release', \E_USER_DEPRECATED);
```

That gives users some more control on this error, e.g. completely ignoring `E_USER_DEPRECATED` messages.

scheb commented on src/ApiClient.php:
```diff
@@ -52,35 +57,72 @@ public function search(string $searchTerm): array
 
     /**
-     * Get historical data for a symbol.
+     * Get historical data for a symbol (depreciated).
```
I believe you wanted to say "deprecated" here

scheb commented on src/ApiClient.php:
```diff
@@ -139,4 +181,43 @@ private function fetchQuotes(array $symbols): array
         return $this->resultDecoder->transformQuotes($responseBody);
     }
+
+    /**
+     * @return string
+     */
```
This can be removed, the method signature is already fully typed.

scheb commented on src/ApiClient.php:
```diff
@@ -139,4 +181,43 @@ private function fetchQuotes(array $symbols): array
+    /**
+     * @return string
+     */
+    private function getResponseBody(string $symbol, string $interval, \DateTimeInterface $startDate, \DateTimeInterface $endDate, string $filter): string
```
Would suggest `getHistoricalDataResponseBody` to be more precise what the method is for.

scheb commented on src/ApiClient.php:
```diff
@@ -139,4 +181,43 @@ private function fetchQuotes(array $symbols): array
+        return (string) $this->client->request('GET', $dataUrl, ['cookies' => $cookieJar])->getBody();
+    }
+
+    /**
+     * @return void
+     * @throws \InvalidArgumentException
+     */
```
I'd remove this PHPDoc (and the one below) as they're pretty much redundant.

scheb commented on src/ResultDecoder.php:
```diff
@@ -135,31 +139,39 @@ public function extractCrumb(string $responseBody): string
-        if (7 !== \count($columns)) {
-            throw new ApiException('CSV did not contain correct number of columns', ApiException::INVALID_RESPONSE);
-        }
-
         try {
-            $date = new \DateTime($columns[0], new \DateTimeZone('UTC'));
+            return new \DateTime($columns[0], new \DateTimeZone('UTC'));
         } catch (\Exception $e) {
             throw new ApiException(sprintf('Not a date in column "Date":%s', $columns[0]), ApiException::INVALID_VALUE);
         }
+    }
```
I don't like so much how this is bound to `$column[0]`. I'd say let's change this to `validateDate(string $value): \DateTime` instead and let the calling method decide what kind of value to pass.

hackeresq committed:
cleanup per discussion

scheb commented on src/ResultDecoder.php:
```diff
@@ -135,31 +139,39 @@ public function extractCrumb(string $responseBody): string
 
     private function createHistoricalData(array $columns): HistoricalData
     {
-        if (7 !== \count($columns)) {
-            throw new ApiException('CSV did not contain correct number of columns', ApiException::INVALID_RESPONSE);
-        }
```
This part is missing in your implementation, that's why tests are failing.

hackeresq replied:
Is that necessary if we're already validating the headers match with `validateHeaderLines()` ?

scheb replied:
I think so. The tests are failing with an "Undefined offset" error, so I'd say we need this check.

scheb commented:
Looks good so far. I figured out why the tests are failing. I'd suggest to add something like that in the new methods as well.

We'd also need test cases for the newly introduced methods.

hackeresq commented:
> We'd also need test cases for the newly introduced methods.

I am very new to tests. And I'm struggling to figure out how you're mocking data. Would love to help write the tests if you could give me a nudge in the right direction?

scheb commented:
So effectively all public methods should be covered by a test case. For the `ResultDecoder` class it's following the principle "put raw data in, check that the output is what you'd expect". Have a look at https://github.com/scheb/yahoo-finance-api/blob/4.x/tests/ResultDecoderTest.php

`transformDividendDataResult` and `transformSplitDataResult` have been added and therefore need test cases. Have a look how `transformHistoricalDataResult` is tested, the same approach should for for these new methods. Should be way simpler though, since it's less data.

hackeresq committed:
removes redundant test

column headers (and thus counts are already checked for expected column headers)

Existing test already checks for this too

hackeresq committed:
adds new mock data

hackeresq committed:
adds tests for new methods

hackeresq committed:
removes pesky trailing comma for 7.1.3 compability

hackeresq committed:
removes another pesky trailing comma for 7.1.3 compability

scheb commented:
Looks good, especially for someone who doesn't have much experience with tests :)

Would you please run `php-cs-fixer fix` once again to fix the code style issues, then I believe we're ready to merge.

hackeresq committed:
runs php-cs-fixer fix

hackeresq commented:
Awesome. Looks like that worked!

scheb commented:
I removed the `$interval` parameter as it turned out to have no effect on the data retrieved.

Released as v4.2.0.
